_posts/2021/2021-09-28-cpp-lessons-learned.md

```diff
@@ -0,0 +1,143 @@
+re-allocated. In Figure 2. below, you can see that all the data is held together:
+metadata, key (right next to `inline_string` structure), and value.
+
+<figure class="image">
+  <img src="/assets/cpp_lessons_inline_string.png" alt="Radix tree with inline_string">
+  <figcaption>Figure 2. Radix_tree with inline_string - only one allocation.</figcaption>
+</figure>
+
+
+* **Using per-thread data**
+
+Very recently, we've described ([in the "Concurrency considerations" blog post][concurrency_blog])
+our approach to ensure data consistency. It's a very challenging task, especially if
+you consider concurrent modifications within a container full of records, with multiple
+threads accessing/modifying the same structure. Read the whole story for details,
+but long story short - per-thread data (implemented in libpmemobj-cpp as
+[enumerable_thread_specific][ptls_docs]) can save you from data consistency issues.
+
+
+* **Keeping locks on pmem is not a good idea**
+
+Since we keep all data for our application on persistent memory, one might think we should
+do that as well with locks. That's the wrong way. Each lock and unlock operation might end up
+with write to pmem. We measured the performance, and it turned out, this approach is slow
+and inefficient. There are two solutions we propose to fix this problem:
+
+The first one is to keep per-element locks in DRAM instead (pointers to these locks are held
+on pmem). Currently, within our library, it is only possible with `concurrent_hash_map`.
+It's a good approach if you store large elements.
+
+The second solution is to use sharding instead of per-element locks. Data can be divided
+into multiple single-threaded buckets, each protected by a lock. With a uniform distribution
+of elements and a big enough number of buckets, this solution will provide a short waiting
+time on locks and good performance. In practice, this can be useful, most likely only in
+unsorted data structures.
+
+
+* **Lazy rehashing**
+
+Usually, in a structure with hashing table, it is required to perform, from time to time,
+a full rehashing of existing entries. In a concurrent environment, this task is very
+challenging and may require locking the entire structure. In our `concurrent_hash_map`
+we introduced a "lazy rehashing" solution, based on the paper
+["Per-bucket concurrent rehashing algorithms"][rehashing_pdf]. When rehashing is required,
+the number of buckets is doubled - some elements may be moved into new buckets, and some
+may stay in the old ones. No locks on buckets are taken, so the whole process happens "lazily"
+in the background. All the algorithm's details can be found in the linked article.
+In short - we use more memory but gain the performance. **Without this approach**,
+the structure (or rather its performance) is vulnerable in moments of mentioned
+full rehashing. These time slots cause latency spikes, slowing down each new write to
```

Comment:
```
**Without this approach**,
the structure (or rather its performance) is vulnerable in moments of mentioned
full rehashing.
```
In context of applications and data structures, word `vulnerable` is most often use to describe security issues. 
mabye: 
```
With this approach, the structure performance do not suffer during full rehashing.
```